idx/metric_definition.go

```diff
@@ -0,0 +1,546 @@
+package idx
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/binary"
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+	"unsafe"
+
+	"github.com/raintank/schema"
+	log "github.com/sirupsen/logrus"
+	"github.com/tinylib/msgp/msgp"
+)
+
+//go:generate msgp
+
+func init() {
+	msgp.RegisterExtension(95, func() msgp.Extension { return &MetricName{} })
+	msgp.RegisterExtension(90, func() msgp.Extension { return &TagKeyValues{} })
+}
+
+type mType uint8
+
+// MType represented by a uint8
+const (
+	MTypeUndefined mType = iota
+	MTypeGauge
+	MTypeRate
+	MTypeCount
+	MTypeCounter
+	MTypeTimestamp
+)
+
+// MetricName stores the name as a []uintptr to strings interned in an object store.
+// Each word is stored as a separate interned string without the '.'. The whole name
+// can be retrieved by calling MetricName.String()
+type MetricName struct {
+	nodes []uintptr
+}
+
+// Nodes returns the []uintptr of interned string addresses
+// for the MetricName
+func (mn *MetricName) Nodes() []uintptr {
+	return mn.nodes
+}
+
+// String returns the full MetricName as a string
+// using data interned in the object store
+func (mn *MetricName) String() string {
+	if len(mn.nodes) == 0 {
+		return ""
+	}
+
+	bld := strings.Builder{}
```

Comment:
this is the canonical way
```suggestion
	var bld strings.Builder
```

Reply:
I would also like to implement a pool of strings.Builder at some point.